modules/exadsBidAdapter.js

```diff
@@ -0,0 +1,435 @@
+
+    const bannerInfo = utils.deepAccess(bidRq.imp[0], 'banner');
+    const nativeInfo = utils.deepAccess(bidRq.imp[0], 'native');
+    const videoInfo = utils.deepAccess(bidRq.imp[0], 'video');
+
+    let w; let h = 0;
+    let mediaType = '';
+    const native = {};
+
+    if (bannerInfo != null) {
+      w = bidRq.imp[0].banner.w;
+      h = bidRq.imp[0].banner.h;
+      mediaType = BANNER;
+    } else if (nativeInfo != null) {
+      const reqNative = JSON.parse(bidResponseAd);
+      reqNative.native.assets.forEach(asset => {
+        if (asset.img != null) {
+          const imgAsset = JSON.parse(bidRq.imp[0].native.request)
+            .native.assets.filter(asset => asset.img != null).map(asset => asset.img);
+          w = imgAsset[0].w;
+          h = imgAsset[0].h;
+          native.image = {
+            url: asset.img.url,
+            height: h,
+            width: w
+          }
+        } else if (asset.title != null) {
+          native.title = asset.title.text;
+        } else if (asset.data != null) {
+          native.body = asset.data.value;
+        }
+      });
+      mediaType = NATIVE;
+    } else if (videoInfo != null) {
+      mediaType = VIDEO;
+    }
+
+    const bidResponse = {
+      requestId: requestId,
+      currency: currency,
+      ad: bidData.adm,
+      cpm: bidData.price,
+      creativeId: bidData.crid,
+      cid: bidData.cid,
+      width: w,
+      ttl: 360,
+      height: h,
+      netRevenue: true,
+      mediaType: mediaType,
+      nurl: pixelUrl
+    };
+
+    if (mediaType == 'native') {
+      native.clickUrl = bidData.adomain[0];
+      bidResponse.native = native;
+    }
+
+    if (mediaType == 'video') {
+      bidResponse.vastXml = bidData.adm;
+      bidResponse.width = bidData.w;
+      bidResponse.height = bidData.h;
+    }
+
+    utils.logInfo('bidResponse->', bidResponse);
+
+    bidResponses.push(bidResponse);
+  } else {
+    utils.logInfo('NO Ad server response ->', serverResponse.body.id);
+  }
+
+  utils.logInfo('interpretResponse -> bidResponses:', bidResponses);
+
+  return bidResponses;
+}
+
+function makeBidRequest(url, data) {
+  const payloadString = JSON.stringify(data);
+
+  return {
+    method: 'POST',
+    url: url,
+    data: payloadString,
+  }
+}
+
+function getUrl(adPartner, bid) {
+  let endpointUrlMapping = {
+    [PARTNERS.RTB_2_4]: bid.params.endpoint + '?idzone=' + bid.params.zoneId + '&fid=' + bid.params.fid
+  };
+
+  return endpointUrlMapping[adPartner] ? endpointUrlMapping[adPartner] : 'defaultEndpoint';
+}
+
+function manageEnvParams() {
+  envParams.domain = window.location.hostname;
+  envParams.page = window.location.protocol + '//' + window.location.host + window.location.pathname;
+  envParams.lang = navigator.language;
+  if (envParams.lang.indexOf('-') > -1) {
+    envParams.lang = envParams.lang.split('-')[0];
+  }
+  envParams.userAgent = navigator.userAgent;
+  if (navigator.appVersion.indexOf('Win') !== -1) {
```

Comment:
Can we use else if ( ) here for these validations with navigator.appVersion , as we are reassigning the value of envParams.osName depending on this value ?